File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
public:
    class Node{
        public:
        int key;
        int val;
        Node*next;
        Node*prev;
        Node(int k,int v){
            key=k;
            val=v;
            next=nullptr;
            prev=nullptr;
        }
    };
    Node*head;
    Node*tail;
    unordered_map<int,Node*>mpp;
    int capacity;
    LRUCache(int capacity) {
        this->capacity=capacity;
        head=new Node(-1,-1);
        tail=new Node(-1,-1);
        head->next=tail;
        tail->prev=head;
    }
    void remove(Node*node){
        Node*prevNode=node->prev;
        Node*nextNode=node->next;
        prevNode->next=nextNode;
        nextNode->prev=prevNode;
    }
    void insert(Node*node){
        Node*nextNode=head->next;
        head->next=node;
        node->prev=head;
        node->next=nextNode;
        nextNode->prev=node;
    }
    int get(int key) {
        if(mpp.find(key)==mpp.end())return -1; 
        Node*existing=mpp[key];
        remove(existing);
        insert(existing);
        return existing->val;
    } 
    void put(int key, int value) {
        Node*node=new Node(key,value);
        if(mpp.find(key)!=mpp.end()){
            Node*existing=mpp[key];
            remove(existing);
            mpp.erase(key);
        }
        if(capacity==mpp.size()){
            Node*lru=tail->prev;
            remove(lru);
            mpp.erase(lru->key);
        }
        insert(node);
        mpp[key]=node;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.java
class LRUCache {
    static class Node{
        int key;
        int val;
        Node next;
        Node prev;
        Node(int k,int v){
            key=k;
            val=v;
        }
    }
    Node head;
    Node tail;
    int capacity;
    Map<Integer,Node>mpp;
    public LRUCache(int capacity) {
        head=new Node(-1,-1);
        tail=new Node(-1,-1);
        head.next=tail;
        tail.prev=head;
        this.capacity=capacity;
        mpp=new HashMap<>();
    }
    void addNode(Node node){
        Node nextNode=head.next;
        head.next=node;
        node.prev=head;
        node.next=nextNode;
        nextNode.prev=node;
    }
    void removeNode(Node node){
        Node prevNode=node.prev;
        Node nextNode=node.next;
        prevNode.next=nextNode;
        nextNode.prev=prevNode;
    }
    public int get(int key) {
        if(!mpp.containsKey(key))return -1;
        Node existing=mpp.get(key);
        removeNode(existing);
        addNode(existing);
        return existing.val;

    }
    public void put(int key, int value) {
        if(mpp.containsKey(key)){
            Node existing=mpp.get(key);
            removeNode(existing);
            mpp.remove(key);
        }   
        if(mpp.size()>=capacity){
            Node lru=tail.prev;
            removeNode(lru);
            mpp.remove(lru.key);
        }
        Node node=new Node(key,value);
        addNode(node);
        mpp.put(key,node);
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */